File: 0084_Largest Rectangle in Histogram/LargestRectangleinHistogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        List<Integer> l = new ArrayList<Integer>();
        for(Integer h : heights){
            l.add(h);
        }

        l.add(0);

        int n = l.size();
        Stack<Integer> s = new Stack<>();
        int ans = 0;
        int i = 0;
        while (i < n) {
          if (s.empty() || l.get(i) >= l.get(s.peek())) {
            s.push(i++);
          } else {
            int h = l.get(s.peek()); 
            s.pop();
            int w = s.empty() ? i : i - s.peek() - 1;        
            ans = Math.max(ans, h * w);
          }
        }
        return ans;
      }
    }

File: 0084_Largest Rectangle in Histogram/LargestRectangleinHistogram.cs
public class Solution {
    public int LargestRectangleArea(int[] heights) {
        var ans = 0;
        if(heights == null || heights.Length == 0) return ans;
        var l = new List<int>(heights.Length+1);
        foreach(var v in heights)
        {
            l.Add(v);
        }
        
        l.Add(0);
        
        var stack = new Stack<int>();
        
        var index = 0;
        while(index < l.Count)
        {
            if(!stack.Any() || l[index] >= l[stack.Peek()])
            {
                stack.Push(index++);
            }else{
                var h = l[stack.Peek()];
                stack.Pop();
                var w = !stack.Any() ? index : index - stack.Peek() - 1;
                ans = Math.Max(ans, h * w);
            }
        }
        
        return ans;
    }
}